arch/eventloop_posix.c

```diff
@@ -0,0 +1,693 @@
+        if(first->timerPolicy == UA_TIMER_HANDLE_CYCLEMISS_WITH_BASETIME)
+            first->nextTime = calculateNextTime(nowMonotonic, first->nextTime,
+                                                (UA_DateTime)first->interval);
+        else
+            first->nextTime = nowMonotonic + (UA_DateTime)first->interval;
+    }
+
+    ZIP_INSERT(UA_TimerZip, &el->timerRoot, first, ZIP_RANK(first, zipfields));
+
+    /* Unlock the mutex before dropping into the callback. So that the timer
+     * itself can be edited within the callback. When we return, only the
+     * pointer to el must still exist. */
+    UA_Callback cb = first->callback;
+    void *app = first->application;
+    void *data = first->data;
+    UA_UNLOCK(&el->elMutex);
+    cb(app, data);
+    UA_LOCK(&el->elMutex);
+}
+
+/* Returns the DateTime of the next cylic callback */
+static UA_DateTime
+processTimer(UA_EventLoop *el, UA_DateTime nowMonotonic) {
+    UA_TimerEntry *first = ZIP_MIN(UA_TimerZip, &el->timerRoot);
+    while(first && first->nextTime <= nowMonotonic) {
+
+        processTimerEntry(el, nowMonotonic, first);
+        first = ZIP_MIN(UA_TimerZip, &el->timerRoot);
+    }
+    /* Return the timestamp of the earliest next callback */
+    return (first) ? first->nextTime : UA_INT64_MAX;
+}
+
+static void
+freeTimerEntry(UA_TimerEntry *te, void *data) {
+    UA_free(te);
+}
+
+/* Process and then free registered delayed callbacks */
+static void
+processDelayed(UA_EventLoop *el) {
+    UA_LOCK_ASSERT(&el->elMutex, 1);
+    while(el->delayedCallbacks) {
+        UA_DelayedCallback *dc = el->delayedCallbacks;
+        el->delayedCallbacks = dc->next;
+        /* Delayed Callbacks might have no cb pointer if all we want to do is
+         * free the memory */
+        if(dc->callback) {
+            UA_UNLOCK(&el->elMutex);
+            dc->callback(dc->application, dc->data);
+            UA_LOCK(&el->elMutex);
+        }
+        UA_free(dc);
+    }
+}
+
+/***********************/
+/* EventLoop Lifecycle */
+/***********************/
+
+UA_EventLoop *
+UA_EventLoop_new(const UA_Logger *logger) {
+    UA_EventLoop *el = (UA_EventLoop*)UA_malloc(sizeof(UA_EventLoop));
+    if(!el)
+        return NULL;
+    memset(el, 0, sizeof(UA_EventLoop));
+    UA_LOCK_INIT(&el->elMutex);
+    el->logger = logger;
+    return el;
+}
+
+UA_StatusCode
+UA_EventLoop_delete(UA_EventLoop *el) {
+    UA_LOCK(&el->elMutex);
+
+    /* Check if the EventLoop can be deleted */
+    if(el->state != UA_EVENTLOOPSTATE_STOPPED &&
+       el->state != UA_EVENTLOOPSTATE_FRESH) {
+        UA_LOG_WARNING(el->logger, UA_LOGCATEGORY_EVENTLOOP,
+                       "Cannot delete a running EventLoop");
+        UA_UNLOCK(&el->elMutex);
+        return UA_STATUSCODE_BADINTERNALERROR;
+    }
+
+    /* Deregister and delete all the EventSources */
+    while(el->eventSources) {
+        UA_EventSource *es = el->eventSources;
+        UA_EventLoop_deregisterEventSource(el, es);
```

Comment:
Same as in https://github.com/open62541/open62541/pull/4553/files#r678903699, we do not have a reentrant lock, so `UA_EventLoop_deregisterEventSource` which itself calls LOCK/UNLOCK fails for the already locked mutex of `UA_EventLoop_delete`.